docs/cardano-monads.rst

```diff
@@ -0,0 +1,608 @@
+
+However, FPComplete began to see ``ether`` as a suboptimal solution, and asked
+us to purge its remains. Now we were supposed to remove all our custom classes,
+replacing them with method records. Those records would go into a ``ReaderT``
+and be passed everywhere manually (as opposed to instance search). Instead of
+distinct ``Ether.MonadReader`` constraints, now we had to use ``MonadReader
+ctx``, passing an annoying ``ctx`` parameter everywhere, and placing constraints
+on it. The final transformer stack is just ``ReaderT ModeCtx Production``, and
+not even a newtype on top.
+
+Technically, we're still in the process of migration, as we haven't removed all
+of our custom classes yet. Just to clarify: in this section we'll discuss the
+current transitional state, and it's more painful than what was actually
+proposed by FPComplete.
+
+In other words, there is a compliсation that actually we have several approaches
+combined:
+
+* General approach a la ExecMode
+* Mockable-based constraints
+* Reflection-based constants (used instead of MonadReader for constant data within single execution)
+* Method dictionaries (for supporting SendActions, a primitive from networking)
+* ReaderT with method dictionaries
+
+General Approach
+~~~~~~~~~~~~~~~~
+
+Now our code follows this pattern::
+
+    -- effect definitions
+
+    class MonadBaz
+        baz :: ...
+
+    defaultBaz = ...
+
+    class MonadZaz
+        zaz :: ...
+
+    defaultZaz = ...
+
+
+    -- mode definitions
+
+    data QuuxCtx = ...
+
+    type QuuxMode = ReaderT QuuxCtx Production
+
+    instance HasFoo ModeCtx
+    instance HasBar ModeCtx
+
+    instance MonadBaz QuuxMode
+        baz = defaultBaz
+
+    instance MonadZaz QuuxMode
+        zaz = defaultZaz
+
+
+Pros:
+
+* Good compile-time performance. There's only a single layer of transformers,
+  instance search is quick, no coercions involved.
+
+Cons:
+
+* Bad run-time configurability. (Same as above)
+
+* Boilerplate. Various ``HasFoo`` instances with field lenses, ``MonadBaz``
+  instances to choose method implementations for the current mode. There's also
+  that annoying ``ctx`` parameter.
+
+* Cost of introducing new modes. The approach is inflexible, as introducing a
+  new mode has an extremely high cost (due to boilerplate). Assuming we want to
+  avoid nested ``ReaderT``, adding one more field to the context requires a new
+  mode.
+
+* Lack of inheritance. It's hard to define one mode in terms of another, with
+  only minor changes. Either it becomes hard to maintain consistency, or it
+  becomes hard to do overrides (as happened to ``AuxxMode``).
+
+* Volatile ``runReaderT``. It's difficult to reason about code when
+  ``runReaderT`` might imply something besides supplyng the value of the
+  ``ReaderT`` environment (handling ``MonadReader``). The situation arises
+  because we define instances for ``ReaderT`` without a newtype (and sometimes
+  even with ``{-# OVERLAPPING #-}``).
+
```

Comment:
Another thing which maybe should be noticed is that such approach forces us to expose internal details. E. g. if we want to be able to record EKG metrics, we have to explicitly provide access to `EKG.Store` via reader config instead of providing only functions to record these metrics.

Approach with settings allows to hide `EKG.store` in function's closure.

Reply:
Good point, but seems orthogonal to the approach described in this section. The `ReaderT` approach implies having method records in the `ReaderT` environment.